torch.go

```diff
@@ -16,27 +17,36 @@ type Optimizer struct {
 }
 
 // RandN returns a tensor filled with random number
-func RandN(rows, cols int, requireGrad bool) Tensor {
+func RandN(rows, cols int, requireGrad bool) *Tensor {
 	rg := 0
 	if requireGrad {
 		rg = 1
 	}
-	return Tensor{C.RandN(C.int(rows), C.int(cols), C.int(rg))}
+	t := &Tensor{C.RandN(C.int(rows), C.int(cols), C.int(rg))}
```

Comment:
We must return `*Tensor` here to make sure that the tensor is created at heap. 
Otherwise, calling `runtime.GC` will destroy all the objects. The training loop would be panic.

Reply:
I don't quite understand this point. What does"all the objects" refer to?

Reply:
@wangkuiyi 

Following is the original `RandN`

```
func RandN(rows, cols int, requireGrad bool) Tensor {
	rg := 0
	if requireGrad {
		rg = 1
	}
	t := Tensor{C.RandN(C.int(rows), C.int(cols), C.int(rg))}
	runtime.SetFinalizer(&t, func(f *Tensor) {
		f.Close()
                fmt.Printf("Close tensor, %d, %d\n", rows, cols)
	})
	return t
}
```
It's returned by value.



Then, in following training loop:

```
a := torch.RandN(100, 10, true)
opt := torch.NewSGDOpt(0.1, 0, 0, 0, false)
opt.AddParameters([]*torch.Tensor{a})

for i := 0; i < 10; i++ {
        fmt.Println(i)
	b := torch.RandN(10, 100, false)
	pre := torch.MM(b, a)
        runtime.GC()
}
```

I call `runtime.GC()`, and expect to free the memory of tensor b only. But it will free the memory of tensor a and tensor b both. The error log is following:

```log
0
Closed tensor 10, 100
Closed tensor 100, 10
1
libc++abi.dylib: terminating with uncaught exception of type c10::Error: Expected a Tensor of type Variable but found an undefined Tensor for argument #1 'mat2' (checked_cast_variable at ../torch/csrc/autograd/VariableTypeManual.cpp:38)
frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 135 (0x4291f47 in libc10.dylib)
frame #1: torch::autograd::VariableType::unpack(at::Tensor const&, char const*, int) + 374 (0x11228a96 in libtorch_cpu.dylib)
frame #2: torch::autograd::VariableType::mm(at::Tensor const&, at::Tensor const&) + 937 (0x10c60ed9 in libtorch_cpu.dylib)
frame #3: c10::detail::wrap_kernel_functor_unboxed_<c10::detail::WrapRuntimeKernelFunctor_<at::Tensor (*)(at::Tensor const&, at::Tensor const&), at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&> >, at::Tensor (at::Tensor const&, at::Tensor const&)>::call(c10::OperatorKernel*, at::Tensor const&, at::Tensor const&) + 21 (0xeb23615 in libtorch_cpu.dylib)
frame #4: at::Tensor c10::KernelFunction::callUnboxed<at::Tensor, at::Tensor const&, at::Tensor const&>(c10::OperatorHandle const&, at::Tensor const&, at::Tensor const&) const + 165 (0x422c5b5 in libcgotorch.so)
frame #5: at::Tensor c10::Dispatcher::callUnboxedWithDispatchKey<at::Tensor, at::Tensor const&, at::Tensor const&>(c10::OperatorHandle const&, c10::DispatchKey, at::Tensor const&, at::Tensor const&) const + 171 (0x422c2db in libcgotorch.so)
frame #6: at::Tensor c10::Dispatcher::callUnboxed<at::Tensor, at::Tensor const&, at::Tensor const&>(c10::OperatorHandle const&, at::Tensor const&, at::Tensor const&) const + 158 (0x422c19e in libcgotorch.so)
frame #7: at::Tensor c10::OperatorHandle::callUnboxed<at::Tensor, at::Tensor const&, at::Tensor const&>(at::Tensor const&, at::Tensor const&) const + 94 (0x422c0ee in libcgotorch.so)
frame #8: at::mm(at::Tensor const&, at::Tensor const&) + 154 (0x4217c9a in libcgotorch.so)
frame #9: MM + 33 (0x4217b91 in libcgotorch.so)
frame #10: _cgo_bdf0e44227b3_Cfunc_MM + 33 (0x40a25b1 in sgd)
frame #11: runtime.asmcgocall + 112 (0x405cec0 in sgd)
```

Reply:
Thanks to @shendiaomo 's help, let me clarify it more clear.

Here is original RandN code:
```go
func RandN(rows, cols int, requireGrad bool) Tensor {
	rg := 0
	if requireGrad {
		rg = 1
	}
	t := Tensor{C.RandN(C.int(rows), C.int(cols), C.int(rg))}
	runtime.SetFinalizer(&t, func(f *Tensor) {
		f.Close()
                fmt.Printf("Close tensor, %d, %d\n", rows, cols)
	})
	return t
}

a := RandN(100, 10, false)
```
`t` is created at the stack, after calling `runtime.GC`, `SetFinalizer` will be called, and the memory is released.
`a` has the same value as `t`. However, the memory is released already, `C.Tensor` of `a` become a dangling pointer.

I made a change:

```go
func RandN(rows, cols int, requireGrad bool) *Tensor {
	rg := 0
	if requireGrad {
		rg = 1
	}
	t := &Tensor{C.RandN(C.int(rows), C.int(cols), C.int(rg))}
	runtime.SetFinalizer(&t, func(f *Tensor) {
		f.Close()
                fmt.Printf("Close tensor, %d, %d\n", rows, cols)
	})
	return t
}

a := RandN(100, 10, false)
```

This time, `t` is created at heap. `a` is a pointer to the Tensor. When calling `runtime.GC`, the lifetime of `a` does not end. It will not be released. The behavior is expected.